README.md
# Security

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/security/workflows/CI/badge.svg)](https://github.com/originphp/security/actions)
[![coverage](https://coveralls.io/repos/github/originphp/security/badge.svg?branch=master)](https://coveralls.io/github/originphp/security?branch=master)

The `Security` library provides various function related to security such as hashing, encryption, decryption and secure random string generation.

## Installation

To install this package

```linux
$ composer require originphp/security
```

## Hashing

### Hashing Strings (Not Passwords)

The default hashing algorithm used is `sha256`.

To hash a string (not password)

```php
use Origin\Security\Security;
$hashed = Security::hash('foo');
```

To hash a string with a pepper (also incorrectly known as salt).

```php
$hashed = Security::hash('foo',['pepper'=>'A random string']);
```

To change the hashing type

```php
$hashed = Security::hash('foo',['type'=>'sha1']);
```

For a full list of supported algorithms

```php
$list = hash_algos()
```

### Hashing Passwords

The Security utility hashes passwords using best practices, currently this is `bcrypt` which is considered
very secure.

```php
$hashed = Security::hashPassword('secret');
```

To verify the password is correct

```php
$hashed = Security::hashPassword('secret');
$bool = Security::verifyPassword($input,$hashed); // input is user inputted password
```

## Encryption

### Generating a Secure Key

To encrypt and decrypt a string you will need a key, you can generate a random secure key.

```php
use Origin\Security\Security;
$key = Security::generateKey(); // ESaCestIJvuAo3NUAtHAZG9DqmFJZtyx
```

The key length must be 32 bytes (256 bits) to use with the encryption decryption functions.

### Encrypting Text

To encrypt a string

```php
use Origin\Security\Security;
$key = '33d80476167cc95c363bf7df3c95e1d1';
$encrypted = Security::encrypt('foo',$key);
```

### Decrypting Text

To decrypt an encrypted string

```php
use Origin\Security\Security;
$key = '33d80476167cc95c363bf7df3c95e1d1';
$encrypted = 'ohRRdAydx+4wfOd7Vm+LHmmV9zBH+3r0WLQylyPMPu2RvCjX9FVgoeUBZuLYBTLM4x9NeZX7U0bUvE1bucATSQ==';
$plain = Security::decrypt($encrypted,$key);
```

## Random Strings

To generate a cryptographically secure hexadecimal random string, the default length is 16. (This is an alias for hex)

```php
$randomString = Security::random(); // 5f31ecf661dabb04
```

### Different Encodings

You can generate secure random strings with different encoding.

```php
$hex = Security::hex(); // gpgf67ezotl06wqs
$base36 = Security::base36(); // 13owqvwcgb426rvq
$base58 = Security::base58(); // SyqBFAtGfNxZkZMQ
$base62 = Security::base62(); // oc1eIfAHKWWt5zrO
$base64 = Security::base64(); // v3xsI6O+g6LsuY4+
// url safe
$base64 = Security::base64(16,true); // YPT9rp-i6jqXWCvA
```

## UID

If you need to generate a unique id, and don't need to use a UUID, then the UID method provides a more memory and disk space efficient way when working with unique ids.

> If you are generating a API token or another form of string that a user might need to type in, then use `Security::random` or `Security::uuid` instead since these use lower case characters.

To generate a cryptographically secure unique id (UID) using base62 with a default length of 16.

```php
$uid = Security::uid(); // O64cjBxfz2JPhyCQ
```

### UUID

The Security class can generate both version 4 and version 1 UUIDs.

To generate a random UUID (version 4)

```php
$uid = Security::uuid(); // 38c67382-d3ab-4430-a27e-0c719813c09f
```

For a version 1 UUID, set MAC address to true, this will try to find the MAC address on Linux systems or generate a random one.

```php
$uid = Security::uuid(['macAddress'=>true]); // ac337932-e4e5-11e9-928f-8bda39fe8887
```

You can also set the MAC address manually.

```php
$uid = Security::uuid(['macAddress'=>'00:0a:95:9d:68:16']); // 769c6fa4-e4e5-11e9-b8d5-000a959d6816
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-01-04

### Changed

- Changed minimum PHP version to 7.3
- Change minimum PHPUnit to 9.2

## [1.1.0] - 2020-07-21

### Changed

- Changed generateKey to use base62

## [1.0.3] - 2020-07-21

### Fixed

- Fixed operators

## [1.0.2] - 2020-05-16

## [1.0.1] - 2019-10-20

### Fixed

- Fixed Mac address return value

## [1.0.0] - 2019-10-16

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).

tests/SecurityTest.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */

namespace Origin\Test\Security;

use Origin\Security\Security;
use \InvalidArgumentException;

class SecurityTest extends \PHPUnit\Framework\TestCase
{
    public function testHash()
    {
        $plain = 'The quick brown fox jumps over the lazy dog';
        $expected = '2fd4e1c67a2d28fced849ee1bb76e7391b93eb12';
       
        $this->assertEquals($expected, Security::hash($plain, ['type' => 'sha1']));
   
        $expected = 'd7a8fbb307d7809469ca9abcb0082e4f8d5651e46d3cdb762d02d0bf37c9e592';
        $this->assertEquals($expected, Security::hash($plain));
        
        $expected = '2a70c8107928b49f2c2b64bac4aacb820aef818b';
        $this->assertEquals($expected, Security::hash($plain, ['type' => 'sha1','pepper' => 'OriginPHP']));

        $this->expectException(InvalidArgumentException::class);
        Security::hash($plain, ['type' => 'unkownHashType']);
    }

    public function testMacAddress()
    {
        if (strtoupper(php_uname('s')) !== 'LINUX') {
            $this->markTestSkipped('This test is for Linux');
        }
        $this->assertRegExp('/^([0-9a-f]{2}:){5}[0-9a-f]{2}$/', Security::macAddress());
    }

    public function testHashPassword()
    {
        $result = Security::hashPassword('secret');
        $this->assertStringContainsString('$2y$10', $result);
    }

    /**
     * @depends testHashPassword
     */
    public function testVerifyPassword()
    {
        $result = Security::hashPassword('secret');
        $this->assertTrue(Security::verifyPassword('secret', $result));
    }

    public function testCompare()
    {
        $expected = crypt('12345', '$2a$07$areallylongstringthatwillbeusedasasalt$');
        $correct = crypt('12345', '$2a$07$areallylongstringthatwillbeusedasasalt$');
        $incorrect = crypt('67890', '$2a$07$areallylongstringthatwillbeusedasasalt$');

        $this->assertTrue(Security::compare($expected, $correct));
        $this->assertFalse(Security::compare($expected, $incorrect));
        $this->assertFalse(Security::compare(null, ''));
    }

    public function testEncryptDecrypt()
    {
        $plain = 'The quick brown fox jumps over the lazy dog';
        $key = '58024d70eb647a3d0654d5211af2ebfd';
  
        $encrypted = Security::encrypt($plain, $key);
        $decrypted = Security::decrypt($encrypted, $key);
        $this->assertEquals($plain, $decrypted);
        $this->assertNull(Security::decrypt($encrypted, str_replace('7', 'a', $key))); // test wrong key

        $this->expectException(InvalidArgumentException::class);
        Security::decrypt($encrypted, $key.'x');
    }

    public function testEncryptInvalidKey()
    {
        $this->expectException(InvalidArgumentException::class);
        Security::encrypt('foo', 'secret');
    }

    public function testDecryptInvalidKey()
    {
        $key = '58024d70eb647a3d0654d5211af2ebfd';

        $encrypted = Security::encrypt('foo', $key);
        $this->expectException(InvalidArgumentException::class);
        Security::decrypt($encrypted, 'secret');
    }

    public function testGenerateKey()
    {
        $this->assertRegExp('/^[a-zA-Z0-9]{32}+$/', Security::generateKey());
    }

    public function testEncryptInvalidKeyLength()
    {
        $this->expectException(InvalidArgumentException::class);
        Security::encrypt('text', 'invalidkey');
    }

    public function testDecryptInvalidKeyLength()
    {
        $this->expectException(InvalidArgumentException::class);
        Security::decrypt('text', 'invalidkey');
    }

    public function testUUID()
    {
        $this->assertRegExp(
            '/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/',
            Security::uuid()
        );

        $this->assertRegExp(
            '/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/',
            Security::uuid(['timestamp' => true])
        );

        $this->assertRegExp(
            '/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/',
            Security::uuid(['macAddress' => true])
        );

        $this->assertRegExp(
            '/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/',
            Security::uuid(['macAddress' => '00:01:01:00:00:00'])
        );

        $this->expectException(InvalidArgumentException::class);
        Security::uuid(['macAddress' => 'example.com']);
    }

    public function testUid()
    {
        $this->assertRegExp('/^[a-zA-Z0-9]{16}$/', Security::uid());
        $this->assertRegExp('/^[a-zA-Z0-9]{18}$/', Security::uid(18));
    }

    public function testRandom()
    {
        $this->assertRegExp('/^[a-f0-9]{16}$/', Security::random());
        $this->assertRegExp('/^[a-f0-9]{21}$/', Security::random(21));
    }

    public function testHex()
    {
        $this->assertRegExp('/^[a-f0-9]{16}$/', Security::hex());
        $this->assertRegExp('/^[a-f0-9]{32}$/', Security::hex(32));
    }
    public function testBase64()
    {
        $regex = '/^[A-Za-z0-9+\/]';
        $this->assertRegExp($regex . '{16}$/', Security::base64());
        $this->assertRegExp($regex . '{32}$/', Security::base64(32));

        $regex = '/^[-A-Za-z0-9_-]';
        $this->assertRegExp($regex . '{64}$/', Security::base64(64, true));
    }

    public function testBase62()
    {
        $chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        $this->assertRegExp('/^[' . $chars . ']{64}$/', Security::base62(64));
    }

    public function testBase58()
    {
        $chars = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
        $this->assertRegExp('/^[' . $chars . ']{64}$/', Security::base58(64));
    }

    public function testBase36()
    {
        $chars = '0123456789abcdefghijklmnopqrstuvwxyz';
        $this->assertRegExp('/^[' . $chars . ']{64}$/', Security::base36(64));
    }
}


src/Security.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types = 1);
namespace Origin\Security;

use InvalidArgumentException;

class Security
{
    const CIPHER = 'AES-256-CBC';

    /**
     * Hashes a string. This is not for passwords.
     *
     * @see https://www.php.net/manual/en/function.hash-algos.php
     * @param string $string
     * @param array $options options keys are
     * - pepper: (default:null). Set a string to use as a pepper
     * - type: (default:sha256) see hash_algos()
     * @return string
     */
    public static function hash(string $string, array $options = []): string
    {
        $options += ['pepper' => null, 'type' => 'sha256'];
       
        if (is_string($options['pepper'])) {
            $string = $options['pepper'] . $string;
        }

        $algorithm = strtolower($options['type']);
        if (! in_array($algorithm, hash_algos())) {
            throw new InvalidArgumentException('Invalid hashing algorithm');
        }

        return hash($algorithm, $string);
    }

    /**
     * Hashes a password using the current best practice which is Bcrypt
     *
     * @param string $password
     * @return string
     */
    public static function hashPassword(string $password): string
    {
        return password_hash($password, PASSWORD_DEFAULT);
    }

    /**
     * Verifies a password against a hash created with hashPassword
     *
     * @param string $password
     * @param string $hash
     * @return bool
     */
    public static function verifyPassword(string $password, string $hash): bool
    {
        return password_verify($password, $hash);
    }

    /**
     * Compares two strings are equal in a secure way to prevent timing attacks.
     *
     * @see https://blog.ircmaxell.com/2014/11/its-all-about-time.html
     *
     * @param string $original
     * @param string $compare
     * @return bool
     */
    public static function compare(string $original = null, string $compare = null): bool
    {
        if (! is_string($original) || ! is_string($compare)) {
            return false;
        }

        return hash_equals($original, $compare);
    }

    /**
     * Generates a secure 256 bits (32 bytes) key
     *
     * The size needs to be the same length as the cipher, anything larger will be truncated.
     *
     * @return string
     */
    public static function generateKey(): string
    {
        return static::base62(32);
    }

    /**
     * Encrypts a string using your key. The key should be secure use. generateKey
     *
     * @see http://php.net/manual/en/function.openssl-encrypt.php
     * @param string $string
     * @param string $key must must be 256 bits (32 bytes)
     * @return string
     */
    public static function encrypt(string $string, string $key) : string
    {
        if (mb_strlen($key) !== 32) {
            throw new InvalidArgumentException('Invalid Key. Key must be 256 bits (32 bytes)');
        }
        $length = openssl_cipher_iv_length(self::CIPHER);
        $iv = random_bytes($length);
        $raw = openssl_encrypt($string, self::CIPHER, $key, OPENSSL_RAW_DATA, $iv);
        $hmac = hash_hmac('sha256', $raw, $key, true);

        return base64_encode($iv . $hmac . $raw);
    }

    /**
     * Decrypts an encrypted string
     *
     * @param string $string
     * @param string $key must must be 256 bits (32 bytes)
     * @return string|null decrypted string
     */
    public static function decrypt(string $string, string $key) : ?string
    {
        if (mb_strlen($key) !== 32) {
            throw new InvalidArgumentException('Invalid Key. Key must be 256 bits (32 bytes)');
        }
        $string = base64_decode($string);
        $length = openssl_cipher_iv_length(self::CIPHER);
        $iv = substr($string, 0, $length);
        $hmac = substr($string, $length, 32);
        $raw = substr($string, $length + 32);
        $expected = hash_hmac('sha256', $raw, $key, true);
       
        if (static::compare($expected, $hmac)) {
            $decrypted = openssl_decrypt($raw, self::CIPHER, $key, OPENSSL_RAW_DATA, $iv);
            if ($decrypted) {
                return $decrypted;
            }
        }

        return null;
    }

    /**
     * Alias for hex (generates a random hex string)
     *
     * @param integer $length 16
     * @return string
     */
    public static function random(int $length = 16) : string
    {
        return static::hex($length);
    }

    /**
     * Generates a cryptographically secure random string that can be used for a unique id.
     *
     * It is designed to be memory & diskspace efficient yet at the same time be unique enough
     * to not have to check the database. This is a solution where you are not required to use a
     * UUID and a user does not need to type.
     *
     * @see https://en.wikipedia.org/wiki/Birthday_problem
     *
     * @param integer $length default 16
     * @return string
     */
    public static function uid(int $length = 16) : string
    {
        return static::base62($length);
    }

    /**
     * Generates a UUID (Universally Unique IDentifier). By default it generates a
     * random UUID (version 4), but if you pass the macAddres option key or provide a
     * MAC address then it will generate a UUID version 1.
     *
     * @param array $options The options array supports the following keys
     *   - macAddress: set to true to use the MAC address (linux) and generate a UUID version 1. You can also set the MAC address manually.
     * @return string
     */
    public static function uuid(array $options = []) : string
    {
        $options += ['macAddress' => null];
        if ($options['macAddress']) {
            if ($options['macAddress'] === true) {
                $options['macAddress'] = static::macAddress() ?: bin2hex(random_bytes(6));
            }

            return static::uuidv1($options['macAddress']);
        }

        return static::uuidv4();
    }

    /**
     * Gets the MAC address on Linux systems which can be used with security features
     *
     * @return string
     */
    public static function macAddress() : ?string
    {
        $macAddress = null;
       
        if (strtoupper(php_uname('s')) === 'LINUX') {
            $files = glob('/sys/class/net/*/address', GLOB_NOSORT);
           
            foreach ($files as $file) {
                $contents = trim(file_get_contents($file));
                if ($contents !== '00:00:00:00:00:00' && preg_match('/^([0-9a-f]{2}:){5}[0-9a-f]{2}$/', $contents)) {
                    $macAddress = $contents;
                    break;
                }
            }
        }
       
        return $macAddress;
    }

    /**
    * Generates a random hex string
    *
    * @param integer $length
    * @return string
    */
    public static function hex(int $length = 16) : string
    {
        $characters = '0123456789abcdef';

        return static::generateRandomString($length, $characters);
    }

    /**
     * Generates a random base36 string
     *
     * @param integer $length
     * @return string
     */
    public static function base36(int $length = 16) : string
    {
        $characters = '0123456789abcdefghijklmnopqrstuvwxyz';

        return static::generateRandomString($length, $characters);
    }

    /**
     * Generates a random base58 string
     *
     * @param integer $length
     * @return string
     */
    public static function base58(int $length = 16) : string
    {
        $characters = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

        return static::generateRandomString($length, $characters);
    }

    /**
    * Generates a random base62 string
    *
    * @param integer $length
    * @return string
    */
    public static function base62(int $length = 16) : string
    {
        $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';

        return static::generateRandomString($length, $characters);
    }

    /**
    * Generates a random base64 string
    *
    * @param integer $length
    * @param boolean $urlSafe use -_ instead of +/
    * @return string
    */
    public static function base64(int $length = 16, bool $urlSafe = false) : string
    {
        $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        $characters .= ($urlSafe === false) ? '+/' : '-_';

        return static::generateRandomString($length, $characters);
    }

    /**
     * Generates a UUID version 4 - Random
     *
     * @return string
     */
    private static function uuidv4() : string
    {
        return implode('-', [
            bin2hex(random_bytes(4)),
            bin2hex(random_bytes(2)),
            bin2hex(chr((ord(random_bytes(1)) & 0x0F) | 0x40)) . bin2hex(random_bytes(1)),
            bin2hex(chr((ord(random_bytes(1)) & 0x3F) | 0x80)) . bin2hex(random_bytes(1)),
            bin2hex(random_bytes(6)),
        ]);
    }

    /**
     * Generates a UUID version 1 - Sequential/Timestamp based
     *
     * @param string $macAddress e.g. 00:0a:95:9d:68:16
     * @return string
     */
    private static function uuidv1(string $macAddress) : string
    {
        $macAddress = str_replace([':', '-'], '', $macAddress);
        if (strlen($macAddress) !== 12 || ctype_xdigit($macAddress) === false) {
            throw new InvalidArgumentException('Invalid MAC address');
        }

        $sequence = random_int(0, 0x3fff); // Not using a stable storage. (RFC 4122 - Section 4.2.1.1)
      
        // Calculate time
        $time = gettimeofday();
        $uuidTime = ($time['sec'] * 10000000) + ($time['usec'] * 10) + 0x01b21dd213814000;

        $timeLow = sprintf('%08x', $uuidTime & 0xffffffff);
        $timeMid = sprintf('%04x', ($uuidTime >> 32) & 0xffff);
        $timeHi = sprintf('%04x', ($uuidTime >> 48) & 0x0fff);
     
        // Apply Version
        $timeHi = hexdec($timeHi) & 0x0fff;
        $timeHi &= ~(0xf000);
        $timeHi |= 1 << 12;

        $sequenceHi = $sequence >> 8;
        $sequenceHi = $sequenceHi & 0x3f;
        $sequenceHi &= ~(0xc0);
        $sequenceHi |= 0x80;

        return vsprintf('%08s-%04s-%04s-%02s%02s-%012s', [
            $timeLow,
            $timeMid,
            sprintf('%04x', $timeHi),
            sprintf('%02x', $sequenceHi),
            sprintf('%02x', $sequence & 0xff),
            $macAddress
        ]);
    }

    /**
     * Random string generator
     *
     * @param integer $length
     * @param string $characters
     * @return string
     */
    private static function generateRandomString(int $length, string  $characters) : string
    {
        $max = strlen($characters) - 1;
        $out = '';
        for ($i = 0; $i < $length; $i++) {
            $out .= $characters[random_int(0, $max)];
        }

        return $out;
    }
}
